shell/platform/darwin/ios/framework/Source/FlutterMetalLayer.mm

```diff
@@ -0,0 +1,388 @@
+  double minFrameRate = fmax(maxFrameRate / 2, 60);
+  if (@available(iOS 15.0, *)) {
+    _displayLink.preferredFrameRateRange =
+        CAFrameRateRangeMake(forceMax ? maxFrameRate : minFrameRate, maxFrameRate, maxFrameRate);
+  } else {
+    _displayLink.preferredFramesPerSecond = maxFrameRate;
+  }
+}
+
+- (void)onDisplayLink:(CADisplayLink*)link {
+  _didSetContentsDuringThisDisplayLinkPeriod = NO;
+  // Do not pause immediately, this seems to prevent 120hz while touching.
+  if (_displayLinkPauseCountdown == 3) {
+    _displayLink.paused = YES;
+    if (_displayLinkForcedMaxRate) {
+      [self setMaxRefreshRate:[DisplayLinkManager displayRefreshRate] forceMax:NO];
+      _displayLinkForcedMaxRate = NO;
+    }
+  } else {
+    ++_displayLinkPauseCountdown;
+  }
+}
+
+- (BOOL)isKindOfClass:(Class)aClass {
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunguarded-availability-new"
+  // Pretend that we're a CAMetalLayer so that the rest of Flutter plays along
+  if ([aClass isEqual:[CAMetalLayer class]]) {
+    return YES;
+  }
+#pragma clang diagnostic pop
+  return [super isKindOfClass:aClass];
+}
+
+- (void)setDrawableSize:(CGSize)drawableSize {
+  [_availableTextures removeAllObjects];
+  _front = nil;
+  _totalTextures = 0;
+  _drawableSize = drawableSize;
+}
+
+- (void)didEnterBackground:(id)notification {
+  [_availableTextures removeAllObjects];
+  _totalTextures = _front != nil ? 1 : 0;
+  _displayLink.paused = YES;
+}
+
+- (CGSize)drawableSize {
+  return _drawableSize;
+}
+
+- (IOSurface*)createIOSurface {
+  unsigned pixelFormat;
+  unsigned bytesPerElement;
+  if (self.pixelFormat == MTLPixelFormatRGBA16Float) {
+    pixelFormat = kCVPixelFormatType_64RGBAHalf;
+    bytesPerElement = 8;
+  } else {
+    pixelFormat = kCVPixelFormatType_32BGRA;
+    bytesPerElement = 4;
+  }
+  size_t bytesPerRow =
+      IOSurfaceAlignProperty(kIOSurfaceBytesPerRow, _drawableSize.width * bytesPerElement);
+  size_t totalBytes =
+      IOSurfaceAlignProperty(kIOSurfaceAllocSize, _drawableSize.height * bytesPerRow);
+  NSDictionary* options = @{
+    (id)kIOSurfaceWidth : @(_drawableSize.width),
+    (id)kIOSurfaceHeight : @(_drawableSize.height),
+    (id)kIOSurfacePixelFormat : @(pixelFormat),
+    (id)kIOSurfaceBytesPerElement : @(bytesPerElement),
+    (id)kIOSurfaceBytesPerRow : @(bytesPerRow),
+    (id)kIOSurfaceAllocSize : @(totalBytes),
+  };
+
+  IOSurfaceRef res = IOSurfaceCreate((CFDictionaryRef)options);
+  if (self.colorspace != nil) {
+    CFStringRef name = CGColorSpaceGetName(self.colorspace);
+    IOSurfaceSetValue(res, CFSTR("IOSurfaceColorSpace"), name);
+  } else {
+    IOSurfaceSetValue(res, CFSTR("IOSurfaceColorSpace"), kCGColorSpaceSRGB);
+  }
+  return (__bridge_transfer IOSurface*)res;
+}
+
+- (FlutterTexture*)nextTexture {
+  @synchronized(self) {
+    if (_totalTextures < 3) {
+      ++_totalTextures;
+      IOSurface* surface = [self createIOSurface];
+      MTLTextureDescriptor* textureDescriptor =
```

Comment:
Does this descriptor benefit from lossy texture compresion and/or do we need to worry about setting MTLTextureUsageShaderRead regressing performance at all? We don't try to read from the framebuffer anymore, but I think with platform views that may still happen.

Reply:
TBH I'm not sure how much of it actually applies for textures created from IOSurfaces. I went with something that looked like a safe default. Btw. does skia not require `MTLTextureUsageShaderRead`?

Reply:
Btw. The texture that backs `CAMetalLayer` drawables also has `MTLTextureUsageShaderRead` set.

Reply:
Its fine to leave it as defaults.

>  Btw. does skia not require MTLTextureUsageShaderRead?

I'm not sure, does this correspond to the framebufferOnly setting on the CAMetalLayer? https://github.com/flutter/engine/blob/7cba4225b9d945b95670b10d21d9e4e1abcff626/shell/platform/darwin/ios/ios_surface_metal_skia.mm#L59

It looks like we need it (though backdrop filters won't read from the framebuffer, i guess platform views might?)

Reply:
Right. Documentation says that with `framebufferOnly` the CALayer textures will only have rendertarget usage. I'll look into it.

I don't see how platform views would read from it. They don't have access to the metal texture, or even the underlying iOSurface, only compositor does.

Reply:
And actually seems necessary, otherwise:
```
[MTLDebugBlitCommandEncoder internalValidateCopyFromTexture:sourceSlice:sourceLevel:sourceOrigin:sourceSize:toTexture:destinationSlice:destinationLevel:destinationOrigin:options:]:388: failed assertion `Copy From Texture Validation
destinationTexture must not be a framebufferOnly texture.
```

Reply:
Interesting, so we can't even blit to the framebuffer only texture! That is good to know. If we ever get back to fixing it we'd have to look into switching to a draw to see if that works